text/0033-input-versions.md

```diff
@@ -0,0 +1,209 @@
+    "contents": [
+        ...
+    ]
+}
+```
+That is, the `[ ... ]` part above is the same format the parser expects today.
+
+Because version information is optional, the Cedar parser should still accept inputs of the form `[ ... ]`, i.e., a list of entities, assuming they are formatted in accord with the most recent version.
+
+### Schema JSON
+
+The Cedar schema JSON is now defined as a record whose elements are individual namespaces, i.e., 
+```
+{
+    "Name::space::one": {
+        ...
+    },
+    "Name::space::two": {
+        ...
+    }
+}
+```
+We propose to extend this format so that one of the "namespace" elements can be `"@cedar_version"`, and it will be associated with the version string XXX, i.e.,
+```
+{
+    "@cedar_version": XXX,
+    "Name::space::one": {
+        ...
+    },
+    "Name::space::two": {
+        ...
+    }
+}
+```
+This approach is unambiguous, and optional, because the element `@cedar_version` cannot be misinterpreted as a namespace, since `@` is not allowed to appear in a namespace identifier.
+
+### Compatibility
+
+The old, version-free format will still be accepted assuming inputs do not rely on changes to format such as [RFC 20](0020-unique-record-keys.md).
+
+## Drawbacks
+
+Implementing this RFC will add some complexity to the various Cedar parsers. They will now have to accept inputs that have a version annotation, and those that do not. They will have to minimally use that annotation to improve functionality and/or error messages. Doing this requires thinking carefully about the meaning of an annotation (e.g., what span of versions use the same format).
+
+The presence of a version annotation may imply more than this RFC is actually offering, leading to customer confusion or disappointment. For example, customers may expect that the current parser can parse and properly interpret all formats up to the present one, rather than reject those earlier formats (gracefully).
+
+Adding an annotation adds some maintenance burden on customers. When a version annotation is present in a policy file, if the policies are changed to use a newer version of Cedar, the version annotation needs to be changed; it is not hard to forget to do so.
+
+If an input format changes in a way that older syntax is interpreted differently, such as the `appliesTo` in the JSON schema as per the request in [issue 351](https://github.com/cedar-policy/cedar/issues/351), the fact that a version annotation is optional presents a problem: A newer and older parser could both accept the same input, but interpret it differently. If a version annotation was required, such an ambiguous interpretation would not be possible.
+
+## Alternatives
+
+### Do nothing to Cedar; version info can appear in application metadata
+
+Rather than add a version annotation in the Cedar input syntax directly, version information can just be stored in metadata. For example, an application could store Cedar schemas in a document database and use a format like
```

Comment:
@hakanson raises a good point that for a single application, the policies, entities, schema, and template links can be viewed as a cohesive unit that should all use the same version. This suggests that we might want a "config file" for Cedar applications that specifies the version, rather than syntax to add a version to each file independently.